src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.EVP.DigestAlgs.cs

```diff
@@ -14,45 +14,79 @@ internal static partial class Crypto
+        private static partial IntPtr CryptoNative_EvpSha256();
 
-        internal static IntPtr EvpSha256() =>
+        private static IntPtr EvpSha256() =>
             s_evpSha256 != IntPtr.Zero ? s_evpSha256 : (s_evpSha256 = CryptoNative_EvpSha256());
 
         [LibraryImport(Libraries.CryptoNative)]
-        internal static partial IntPtr CryptoNative_EvpSha384();
+        private static partial IntPtr CryptoNative_EvpSha384();
 
-        internal static IntPtr EvpSha384() =>
+        private static IntPtr EvpSha384() =>
             s_evpSha384 != IntPtr.Zero ? s_evpSha384 : (s_evpSha384 = CryptoNative_EvpSha384());
 
         [LibraryImport(Libraries.CryptoNative)]
-        internal static partial IntPtr CryptoNative_EvpSha512();
+        private static partial IntPtr CryptoNative_EvpSha512();
 
-        internal static IntPtr EvpSha512() =>
+        private static IntPtr EvpSha512() =>
             s_evpSha512 != IntPtr.Zero ? s_evpSha512 : (s_evpSha512 = CryptoNative_EvpSha512());
 
-        internal static IntPtr HashAlgorithmToEvp(string hashAlgorithmId) => hashAlgorithmId switch
+        [LibraryImport(Libraries.CryptoNative)]
+        private static partial IntPtr CryptoNative_EvpSha3_256();
+
+        private static IntPtr EvpSha3_256() =>
+            s_evpSha3_256 != -1 ? s_evpSha3_256 : (s_evpSha3_256 = CryptoNative_EvpSha3_256());
+
+        [LibraryImport(Libraries.CryptoNative)]
+        private static partial IntPtr CryptoNative_EvpSha3_384();
+
+        private static IntPtr EvpSha3_384() =>
+            s_evpSha3_384 != -1 ? s_evpSha3_384 : (s_evpSha3_384 = CryptoNative_EvpSha3_384());
+
+        [LibraryImport(Libraries.CryptoNative)]
+        private static partial IntPtr CryptoNative_EvpSha3_512();
+
+        private static IntPtr EvpSha3_512() =>
+            s_evpSha3_512 != -1 ? s_evpSha3_512 : (s_evpSha3_512 = CryptoNative_EvpSha3_512());
+
+        internal static IntPtr HashAlgorithmToEvp(string hashAlgorithmId)
         {
-            nameof(HashAlgorithmName.SHA1) => EvpSha1(),
-            nameof(HashAlgorithmName.SHA256) => EvpSha256(),
-            nameof(HashAlgorithmName.SHA384) => EvpSha384(),
-            nameof(HashAlgorithmName.SHA512) => EvpSha512(),
-            nameof(HashAlgorithmName.MD5) => EvpMd5(),
-            _ => throw new CryptographicException(SR.Format(SR.Cryptography_UnknownHashAlgorithm, hashAlgorithmId))
-        };
+            switch (hashAlgorithmId)
+            {
+                case HashAlgorithmNames.SHA1: return EvpSha1();
+                case HashAlgorithmNames.SHA256: return EvpSha256();
+                case HashAlgorithmNames.SHA384: return EvpSha384();
+                case HashAlgorithmNames.SHA512: return EvpSha512();
+                case HashAlgorithmNames.SHA3_256:
+                    IntPtr sha3_256 = EvpSha3_256();
+                    return sha3_256 != 0 ? sha3_256 : throw new PlatformNotSupportedException();
```

Comment:
Do any of the myriad PNSEs in this change need a custom string, or will it always be pretty obvious that the thing not supported is a SHA-3?

Reply:
Most straight forward uses of SHA3 like `SHA3_256` are going to check is supported in Create or their constructor, etc. Many other places try to guard against unsupported SHA3 (like RSA.SignData) guard as well.

We can change it to have a custom string if you prefer, but the call site that throws should make it pretty obvious. The throw here is unlikely to even get hit since we do more up-front checking, but, I can imagine there might be some path to this that doesn't have a precondition check, so an assert felt wrong.